data-analytics/dataflow-elasticsearch-indexer/ElasticIndexer/src/main/java/com/google/cloud/pso/common/FailSafeValidate.java

```diff
@@ -0,0 +1,147 @@
+
+        abstract Builder setObjectReader(ObjectReader objectReader);
+
+        abstract Builder setExtractKeyFn(ExtractKeyFn extractKeyFn);
+
+        public abstract FailSafeValidate build();
+
+
+        public Builder withKeyPath(String keyPath) {
+            checkArgument(keyPath != null,
+                    "withKeyPath(keyPath) called with null value.");
+            try {
+                JsonPointer.compile(keyPath);
+            } catch (IllegalArgumentException e) {
+                throw new RuntimeException(e);
+            }
+            return setExtractKeyFn(new ExtractKeyFn(keyPath));
+        }
+
+        public Builder withSuccessTag(TupleTag<KV<String, String>> successTag) {
+            checkArgument(successTag != null,
+                    "withSuccessTag(successTag) called with null value.");
+            return setSuccessTag(successTag);
+        }
+
+        public Builder withFailureTag(TupleTag<ErrorMessage> failureTag) {
+            checkArgument(failureTag != null,
+                    "withFailureTag(failureTag) called with null value.");
+            return setFailureTag(failureTag);
+        }
+    }
+
+
+    @Override
+    public PCollectionTuple expand(PCollection<String> input) {
+        return input.apply("ValidateAndTagMessages",
+                ParDo.of(
+                        new DoFn<String, KV<String, String>>() {
+                            @ProcessElement
+                            public void processElement(ProcessContext context) {
+                                String input = context.element();
+
+                                try {
+                                    JsonParser parser = objectReader().getFactory().createParser(input);
+                                    // Parse the json to validate it. An IOException is thrown if
+                                    // it encounters a parsing error.
+                                    while(parser.nextToken() != null) {}
+                                    JsonNode jsonNode = objectReader().readTree(input);
+
+                                    String key = null;
+
+                                    try {
+                                        key = extractKeyFn().apply(jsonNode);
+
+                                    } catch (RuntimeException e) {
+                                        ErrorMessage em = ErrorMessage.newBuilder()
+                                                .withJsonPayload(input)
+                                                .withErrorMessage("Missing key element")
+                                                .build();
+                                        context.output(failureTag(), em);
+                                    }
+
+                                    if (key != null) {
+                                        context.output(successTag(),
+                                                KV.of(key, input));
```

Comment:
Is there another value we can pass onward so we don't have to do JSON processing again the InsertMetadataFn?